Support single page applications

AliceWaddicor opened the pull request:
Support Single page applications by allowing different variants of a single page to be represented by separate page models. 
This was previously not possible as the application would try to reset the current page model, based on the URL, on navigation and click events. If more than one registered page model had the same URL an exception would occur. 
Fix:
Added a single page application flag to the config file.
Prevent the current page being reset (using the URL) on navigation and click events, if the flag is true
Updated the readme with instructions for single page application testing

AliceWaddicor committed:
Add SinglePageApplication flag to config, prevent reset of Navigator.CurrentPage on nav & click events if true

AliceWaddicor committed:
Change config from SinglePageApplication to UseNavigationMap, SPAs with different URLs are supported by adding the route marker

AliceWaddicor committed:
Add SPA functionality to the homepage of the Sample.Web app

AliceWaddicor committed:
Add a new Specs project to test the site as a SPA application with routing

jamesbroome commented on Solutions/Endjin.SpecFlow.Selenium.Framework.Net45/Navigation/NavigationMap.cs:
```diff
@@ -31,6 +31,8 @@ private PageMapEntry HomePage
         }
 
+        public virtual string SpaRoutingMarker => null;
+
```
Shouldn't SpaRoutingMarker be Protected - it's only ever going to be set by the inheriting class?

jamesbroome commented on Solutions/Endjin.SpecFlow.Selenium.Sample.Specs.Spa/Steps/CommonSteps.cs:
```diff
@@ -0,0 +1,28 @@
+
+    [Binding]
+    public class CommonSteps
+    {
```
This class is duplication of CommonSteps.cs in the other specs project.

MikeEvansLarah replied:
We should probably move it down into the framework library

jamesbroome commented on Solutions/Endjin.SpecFlow.Selenium.Framework.Net45/WebsiteUnderTest/ReadMe.txt:
```diff
@@ -32,6 +32,14 @@ this.AddPage<RegisterPageModel>(Page.Register, "Account/Register");
+Setting this option to false allows the library to support single page applications which have a single invariant URL for all views 
+Single page applications which display other route marks (e.g. '#' or '#!') in the URL can use the navigation map
+Single page applications which display other route marks (e.g. '#' or '#!') in the URL should specify the SpaRoutingMarker in the WebsiteNavigationMap. The routing marker should not be included in MapPages() -->
+"use_navigation_map": false
```
Shouldn't this be shown as true to match the default setting in the json config (otherwise a bit confusing)

jamesbroome commented on Solutions/Endjin.SpecFlow.Selenium.Sample.Specs.Spa/WebsiteUnderTest/web_test.config.json:
```diff
@@ -0,0 +1,20 @@
+﻿{
+  "web_app_name": "Endjin.SpecFlow.Selenium.Sample.Web",
+  "website_url": "http://localhost:1727/",
+  "use_navigation_map": true,
```
Shouldn't this be set to false for the Spa sample?

MikeEvansLarah replied:
No, you still use the navigation map for SPA. You just set the SpaRoutingMarker on the map.

jamesbroome commented on Solutions/Endjin.SpecFlow.Selenium.Sample.Specs/WebsiteUnderTest/ReadMe.txt:
```diff
@@ -32,6 +32,14 @@ this.AddPage<RegisterPageModel>(Page.Register, "Account/Register");
+Setting this option to false allows the library to support single page applications which have a single invariant URL for all views 
+Single page applications which display other route marks (e.g. '#' or '#!') in the URL can use the navigation map
+Single page applications which display other route marks (e.g. '#' or '#!') in the URL should specify the SpaRoutingMarker in the WebsiteNavigationMap. The routing marker should not be included in MapPages() -->
+"use_navigation_map": false
```
Shouldn't this be shown as true to match the default setting in the json config (otherwise a bit confusing)

AliceWaddicor committed:
Add .vs/ to gitignore to remove applicationhost.config

AliceWaddicor committed:
Merged 'develop'.

# Conflicts:
#	Solutions/Endjin.SpecFlow.Selenium.Framework.Net45/Navigation/NavigatorSessionParameters.cs
#	Solutions/Endjin.SpecFlow.Selenium.Framework.Net45/WebsiteUnderTest/web_test.config.json
#	Solutions/Endjin.SpecFlow.Selenium.Sample.Specs/WebsiteUnderTest/web_test.config.json

AliceWaddicor committed:
Update NUnit to match other projects in Endjin.SpecFlow.Selenium.Sample.Specs.Spa

AliceWaddicor committed:
Changed accessibility of SpaRoutingMarker to protected

AliceWaddicor committed:
Showed the value of use_navigation_map as true in the readme as this is the more common usage

AliceWaddicor committed:
Updated the web_test.config.json in Selenium.Sample.Specs.Spa to match other projects

AliceWaddicor committed:
Moved step definitions which were in CommonSteps.cs into the library
Having the definitions in the .tt file means the feature files get the correct syntax highlighting

AliceWaddicor committed:
Removed C# 6 language features
